spam/spam_data_augmentation_tutorial.py

```diff
@@ -0,0 +1,274 @@
+        synonym = get_synonym(words[to_replace], pos="n")
+        if synonym and synonym != words[to_replace]:
+            x.text = " ".join(words[:to_replace] + [synonym] + words[1 + to_replace :])
+            return x
+
+
+@transformation_function(pre=[spacy])
+def replace_adjective_with_synonym(x):
+    words = [token.text for token in x.doc]
+    idxs = [i for i, token in enumerate(x.doc) if token.pos_ == "ADJ"]
+    if len(idxs) > 1:
+        to_replace = np.random.choice(idxs[:-1])
+        synonym = get_synonym(words[to_replace], pos="a")
+        if synonym and synonym != words[to_replace]:
+            x.text = " ".join(words[:to_replace] + [synonym] + words[1 + to_replace :])
+            return x
+
+
+# %% [markdown]
+# We can try running the TFs on our training data to demonstrate their effect.
+
+# %%
+tfs = [
+    change_person,
+    swap_nouns,
+    replace_verb_with_synonym,
+    replace_noun_with_synonym,
+    replace_adjective_with_synonym,
+]
+
+for tf in tfs:
+    for i, row in df_train.iterrows():
+        transformed_or_none = tf(row)
+        if transformed_or_none is not None:
+            print(f"TF Name: {tf.name}")
+            print(f"Original Text: {row.text}")
+            print(f"Transformed Text: {transformed_or_none.text}\n")
+            break
+
+# %% [markdown]
+# ## 3. Applying Transformation Functions
+
+# %% [markdown]
+# To apply one or more TFs that we've written to a collection of data points, we use a `TFApplier`.
+# Because our data points are represented with a Pandas DataFrame in this tutorial, we use the `PandasTFApplier` class. In addition, we can apply multiple TFs in a sequence to each example. A `policy` is used to determine what sequence of TFs to apply to each example. In this case, we just use a `RandomPolicy` that picks 3 TFs at random per example. The `n_per_original` argument determines how many augmented examples to generate per original example.
+#
+
+# %%
+from snorkel.augmentation.apply import PandasTFApplier
+from snorkel.augmentation.policy import RandomPolicy
+
+policy = RandomPolicy(len(tfs), sequence_length=2, n_per_original=2, keep_original=True)
+tf_applier = PandasTFApplier(tfs, policy)
+df_train_augmented = tf_applier.apply(df_train)
+Y_train_augmented = df_train_augmented["label"].values
+
+# %%
+print(f"Original training set size: {len(df_train)}")
+print(f"Augmented training set size: {len(df_train_augmented)}")
+
+# %% [markdown]
+# We have almost doubled our dataset using TFs! Note that despite `n_per_original` being set to 2, our dataset may not exactly triple in size, because some TFs keep the example unchanged (e.g. `change_person` when applied to a sentence with no persons).
+
+# %% [markdown]
+# ## 4. Training an End Model
+#
+# Our final step is to use the augmented data to train a model. We train an LSTM (Long Short Term Memory) model, which is a commonly used architecture for text processing tasks.
+
+# %%
+import tensorflow as tf
+
```

Comment:
We should be very clear that the below is *boilerplate* code (I assume it is right?), because to a naive observer.  Even better would be if there is a nice wrapper in keras or another lib that reduces the number of lines needed below... key is just to be clear we're using a generic end model